inluxclient/query.go

```diff
@@ -0,0 +1,315 @@
+package influxclient
+
+import (
+	"encoding/base64"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// TableColumn holds flux query table column properties
+type TableColumn struct {
+	Name         string
+	DataType     string
+	IsGroup      bool
+	DefaultValue string
+}
+
+// String returns TableColumn string dump
+func (f *TableColumn) String() string {
+	return fmt.Sprintf("{name: %s, datatype: %s, defaultValue: %s, group: %v}", f.Name, f.DataType, f.DefaultValue, f.IsGroup)
+}
+
+const (
+	stringDatatype       = "string"
```

Comment:
These constants will be used also in the `Decode()` function, to validate type compatibility.

Reply:
That's quite possible, but the Decode function doesn't exist yet, so it's hard to be sure if actually will be necessary (I can  imagine an implementation where there's still only a single switch statement - if fact I suspect it might be a good idea to implement toValue in terms of Decode, but we'll have to see about that. I'd suggest keeping the code as simple as possible at each step and refactoring when needed. That way it's easy to review and understand each PR on its own terms without guessing the future.
Of course there might be times when we have to look ahead a bit in the implementation, but I'm not sure this is one of those.

Reply:
I have the `Decode` in progress in [my fork](https://github.com/bonitoo-io/influxdb-client-go/blob/tmp/query_cont/inluxclient/query_decode.go). It is far from finishing, but if it is going in the wrong direction,  please, let me know.

Reply:
Thanks for pointing me to that. I had a think about it, and I think I'd choose a slightly different direction (sorry to be a pain again!). Two main things:
- I think that rather than iterating over all the fields of the type, it might be better to iterate over the known columns instead (on the basis that if you're decoding into a struct with fields that aren't mentioned in the columns, that probably doesn't matter).
- It's usually worth trying to avoid doing more type reflection work than necessary - it would be nice to make this reasonably efficient for tables with many rows.

The second point is conventionally addressed by creating a closure to do the work for a given known type, with as many aspects as possible factored out of the inner part of the function. In many places (e.g. encoding/json), that's done with a global `sync.Map` variable keyed by type, but in this case, I think we might be able to assume that tables have enough rows that just caching the converter function for the latest known type and set of columns is enough.

I did a very rough and rapid sketch of how I think the structure of the decoder might look in this style:

```
type Reader struct {
	// decodeType holds the type that was last decoded into.
	// This is reset at the start of each section.
	decodeType reflect.Type
	// colSetters holds an element for each column in the current
	// section. Given an instance of decodeType, it decodes
	// the column string value into the appropriate field.
	colSetters []func(reflect.Value, string) error
}

// initColumns initializes r.colSetters for the given columns.
// It's called once per section.
func (r *Reader) initColumns(t reflect.Type, columns []columnInfo) error {
	if t == r.decodeType {
		return nil
	}
	if t.Kind() != reflect.Ptr {
		return fmt.Errorf("cannot decode into non-pointer value")
	}
	t = t.Elem()
	if t.Kind() != reflect.Struct {
		return fmt.Errorf("cannot decode into non-struct value")
	}
	setters := make([]fieldSetter, len(columns))
	
	for i, col := range columns {
		find corresponding field in struct
		fkind := fieldKind(fieldType)
		convert, ok := conversions[conv{columnType, fkind]
		if !ok {
			error('cannot convert from column type %v to %v", fieldType, columnType)
		}
		setters[i] = func(v reflect.Value, s string) error {
			fv := v.FieldByIndex(fieldIndex)
			return convert(fv, s)
		}
	}
	r.decodeType = t
	return nil
}

// conv represents a possible conversion source and destination.
type conv struct {
	from columnKind
	to reflect.Kind
}

// fieldKind represents one of the possible kinds of struct field.
// It's similar to reflect.Kind (every reflect.Kind constant i
// is represented as fieldKind(i)) except that it also has
// defined constants for field types that have special treatment,
// such as time.Duration.
type fieldKind uint

var intKinds = []fieldKind{
	fieldKind(reflect.Int),
	fieldKind(reflect.Int8),
	fieldKind(reflect.Int16),
	etc,
}

const (
	durationKind = fieldKind(255 + iota)
	timeKind
	bytesKind
)

type colType uint

const (
	stringCol = colType(iota)
	durationCol
	longCol
	etc
)

type fieldSetter = func(v reflect.Value, s string) error

var conversions = func() map[conv] fieldSetter {
	m := make(map[conv] fieldSetter)
	for _, k := range intKinds {
		m[conv{longCol, k}] = toInt
		m[conv{uLongCol, k}] = toInt
	}
	for _, k := range uintKinds {
		m[conv{uLongCol, k}] = toUint
	}
	for _, k := range floatKinds {
		m[conv{longCol, k}] = toFloat
		m[conv{uLongCol, k}] = toFloat
		m[conv{doubleCol, k}] = toFloat
	}
	m[conv{stringCol, reflect.String}] = toString
	m[conv{durationCol, durationKind}] = toDuration
	m[conv{durationCol, reflect.String}] = toString
	m[conv{base64BinaryCol, bytesKind}] = toBytes
	m[conv{timeColRFC, timeKind}] = toTime
	m[conv{timeColRFCNano, timeKind}] = toTime
	
	for col, t := range canonicalKinds {
		m[conv{col, reflect.Interface}] = toInterface(col)
	}
}()

// canonicalTypes holds the Go type that best represents
// all the column kinds (keyed  by column type).
var canonicalTypes = []reflect.Type {
	stringCol: reflect.TypeOf(""),
	durationCol: reflect.TypeOf(time.Duration),
	longCol: reflect.TypeOf(int64(0)),
	uLongCol: reflect.TypeOf(uint64(0))
	doubleCol: reflect.TypeOf(0.0),
	timeColRFC: reflect.TypeOf(time.Time{}),
	timeColRFCNano: reflect.TypeOf(time.Time{}),
}

var basicKinds = map[reflect.Kind] bool {
	reflect.Int,
	reflect.Int16,
	...
	reflect.Uint,
	...
	reflect.

func fieldKindOf(t reflect.Type) fieldKind {
	switch t {
	case time.Duration:
		return durationKind
	case time.Time:
		return timeKind
	case reflect.Slice:
		if reflect.Elem().Kind() == reflect.Uint8 {
			return bytesKind
		}
	}
	return fieldKind(t.Kind())
}

func toInterface(col colType) fieldSetter {
	t := canonicalKinds[col]
	convert, ok := conversions[conv{col, fieldKindOf(t)]
	if !ok {
		panic("conversion not found (should not happen)")
	}
	return func(v reflect.Value, s string) error {
		e := reflect.New(t).Elem()
		if err := convert(e, s); err != nil {
			return err
		}
		v.Set(e)
		return nil
	}
}

func toString(v reflect.Value, s string) error {
	v.SetString(s)
	return nil
}

func toFloat(v reflect.Value, s string) error {
	x, err := strconv.ParseFloat(s, 64, 10)
	if err != nil {
		return err
	}
	if v.OverflowFloat(x) {
		return fmt.Errorf("overflow")
	}
	v.SetFloat(x)
	return nil
}

func toBool(v reflect.Value, s string) error {
	var b bool
	switch s {
	case "true":
		b = true
	case "false"
		b = false
	default:
		return fmt.Errorf("invalid bool value")
	}
	v.SetBool(b)
	return nil
}

func toInt(v reflect.Value, s string) error {
	x, err := strconv.ParseInt(s, 64, 10)
	if err != nil {
		return err
	}
	if v.OverflowInt(x) {
		return fmt.Errorf("overflow")
	}
	v.SetInt(x)
	return nil
}
```

Note: Instead of having string constants for the column types, I've defined a numeric type - there would be some code somewhere to convert to that via a switch or a map lookup.